hydra-node/src/Hydra/Network/Ouroboros.hs

```diff
@@ -129,34 +130,62 @@ withOuroborosNetwork tracer localHost remoteHosts networkCallback between = do
   let newBroadcastChannel = atomically $ dupTChan bchan
   -- TODO: Factor this out, there should be only one IOManager per process.
   withIOManager $ \iomgr -> do
-    race_ (connect iomgr newBroadcastChannel hydraClient) $
+    -- init a signal mvar used by Network.setPeers to communicate with thread 'connect'
+    signal <- newMVar remoteHosts
+    -- init read signal to avoid blocking on Network.getPeers
+    readSignal <- newMVar remoteHosts
+    race_ (connect iomgr newBroadcastChannel hydraClient signal) $
       race_ (listen iomgr hydraServer) $ do
         between $
           Network
             { broadcast = atomically . writeTChan bchan
+            , getPeers = readMVar readSignal
+            , setPeers = \peers -> do
+                putMVar signal peers
+                modifyMVar readSignal peers
             }
  where
+  modifyMVar signal peers = do
+    _ <- takeMVar signal
+    putMVar signal peers
```

Comment:
This is relatively unsafe (not thread safe) to do in a non-atomic manner. If anything, you should use [the built-in from the `Control.Concurrent` module](https://hackage.haskell.org/package/base-4.17.0.0/docs/Control-Concurrent-MVar.html#v:modifyMVar) or better, as suggested above, use [io-sim-classes](https://github.com/input-output-hk/io-sim/blob/main/io-classes/src/Control/Monad/Class/MonadSTM.hs#L120-L143)